scrapy/core/http2/protocol.py

```diff
@@ -0,0 +1,183 @@
+        self.streams = {}
+
+        # Boolean to keep track the connection is made
+        # If requests are received before connection is made
+        # we keep all requests in a pool and send them as the connection
+        # is made
+        self.is_connection_made = False
+        self._pending_request_stream_pool = []
+
+    def _stream_close_cb(self, stream_id: int):
+        """Called when stream is closed completely
+        """
+        self.streams.pop(stream_id, None)
+
+    def _new_stream(self, request: Request):
+        """Instantiates a new Stream object
+        """
+        stream_id = self.next_stream_id
+        self.next_stream_id += 2
+
+        stream = Stream(
+            stream_id=stream_id,
+            request=request,
+            connection=self.conn,
+            write_to_transport=self._write_to_transport,
+            cb_close=self._stream_close_cb
+        )
+
+        self.streams[stream.stream_id] = stream
+        return stream
+
+    def _send_pending_requests(self):
+        # TODO: handle MAX_CONCURRENT_STREAMS
+        # Initiate all pending requests
+        for stream in self._pending_request_stream_pool:
+            stream.initiate_request()
+
+        self._pending_request_stream_pool.clear()
+
+    def _write_to_transport(self):
+        """ Write data to the underlying transport connection
+        from the HTTP2 connection instance if any
+        """
+        data = self.conn.data_to_send()
+        self.transport.write(data)
+
+    def request(self, _request: Request):
+        stream = self._new_stream(_request)
+        d = stream.get_response()
+
+        # If connection is not yet established then add the
+        # stream to pool or initiate request
+        if self.is_connection_made:
+            stream.initiate_request()
+        else:
+            self._pending_request_stream_pool.append(stream)
+
+        return d
+
+    def connectionMade(self):
+        """Called by Twisted when the connection is established. We can start
+        sending some data now: we should open with the connection preamble.
+        """
+        self.destination = self.transport.connector.getDestination()
+        LOGGER.info('Connection made to {}'.format(self.destination))
+
+        self.conn.initiate_connection()
+        self._write_to_transport()
+        self.is_connection_made = True
+
+        self._send_pending_requests()
+
+    def dataReceived(self, data):
+        events = self.conn.receive_data(data)
+        self._handle_events(events)
+        self._write_to_transport()
+
+    def connectionLost(self, reason=connectionDone):
+        """Called by Twisted when the transport connection is lost.
+        No need to write anything to transport here.
+        """
+        # Pop all streams which were pending and were not yet started
+        for stream_id in list(self.streams):
+            self.streams[stream_id].lost_connection()
```

Comment:
I must say things seem a bit overcomplicated. It is weird that `protocol.streams[n].lost_connection()` pops `protocol.streams[n]` (through a callback).

Any reason not to make it simpler?

```suggestion
        for stream in self.streams.values():
            stream.stream_closed_server = True
        self.streams = {}
```

Given the protocol is the one that calls `reset`, `lost_connection` and `end_stream`, the callback approach seems unnecessary, at least given the current code.

Reply:
> Any reason not to make it simpler?

One of the reasons was I wanted to make code more readable and was planning to handle cases where in functins `lost_connection` and `reset` we have to fire the reponse deferred errback. Similarly for `end_stream` we have to fire response deferred callback with the response instance